### PR TITLE
Update linear solver safe mode

### PR DESCRIPTION
**Pull request by cnpetra**
The outer optimization loop switches between "safe" and "fast" KKT formulations when various convergence metrics (or lack of them) occur in the Krylov-based iterative refinement for the GPU Cholesky solves and in the filter line-search algorithm. Builds on the "safe mode" idea used for MDS linear algebra but now it allows switching between KKT formulations on the fly during the optimization.

### CODE REVIEW

**Comment by nychiang** [comment]
Rebase to or merge with `develop` to clear the CI failure on PNNL gitlab

File: src/Optimization/hiopAlgFilterIPM.cpp [hunk 1]
```diff
@@ -1055,7 +1055,7 @@ hiopSolveStatus hiopAlgFilterIPMQuasiNewton::run()
     /************************************************
      * update mu and other parameters
      ************************************************/
-    while(_err_log<=kappa_eps * _mu) {
+    if(_err_log<=kappa_eps * _mu) {
       //update mu and tau (fraction-to-boundary)
       bret = updateLogBarrierParameters(*it_curr, _mu, _tau, _mu, _tau);
       if(!bret) break; //no update is necessary
```
Comment (nychiang) [inline src/Optimization/hiopAlgFilterIPM.cpp:1061]: why do we change this `while` loop to `if`? I think `while` is designed to reduce `mu` faster.
In addition, if we change to `if`,  the `break` command on line 1061 may break the outer loop and give us unpredictable result.

**Comment by cnpetra** [comment]
merged this PR in `csr-interface-dev` and we will review and discuss it there.

File: src/Optimization/hiopAlgFilterIPM.cpp (hunk 1 above)
Comment (cnpetra) [reply src/Optimization/hiopAlgFilterIPM.cpp:1061]: under `elastic_mode` using `while` tighten up bounds too quickly and enters FR quite often. Also, aggressive reduction puts a lot of stress on the linear solver for some reason. Not sure using `while` for the non`elastic_mode` outweigh the, but I can leave the `while` on for nonelastic mode.

Will remove fix the `break`.

**Pull request merged**